### PR TITLE
Gas management simplification

### PR DESCRIPTION
**Pull request by camilbancioiu**
This PR rewrites gas management for stacked SC execution. Previously, caller contracts would "forward" gas to the contracts they called in order to "balance the books" for gas accounting. The gas forwarding mechanism was cumbersome and easy to misuse.

The solution was to redefine the state stack of the `MeteringContext` as a mapping of contract addresses to amounts of gas used by the contracts. Everytime the internal state of `MeteringContext` is initialized, the current mapping is cloned and pushed onto the internal state stack, then reset to 0. This correctly isolates the gas accounting of subsequent contract executions from the previous ones.

### CODE REVIEW

File: arwen/host/execution.go [hunk 1]
```diff
@@ -930,7 +898,7 @@ func (host *vmHost) callSCMethod() error {
 		err = host.handleBreakpointIfAny(err)
 	}
 	if err == nil {
-		err = host.checkFinalGasAfterExit()
+		// err = host.checkFinalGasAfterExit()
```
Comment (sasurobert) [inline arwen/host/execution.go:901]: this is a must :D

File: arwen/host/execution_test.go [hunk 2]
```diff
@@ -1448,7 +1448,7 @@ func TestExecution_CreateNewContract_Fail(t *testing.T) {
 func TestExecution_Mocked_Wasmer_Instances(t *testing.T) {
 	host, _, ibm := defaultTestArwenForCallWithInstanceMocks(t)
 
-	parentInstance := ibm.CreateAndStoreInstanceMock(parentAddress, 1000)
+	parentInstance := ibm.CreateAndStoreInstanceMock(t, host, parentAddress, 1000)
```
Comment (sasurobert) [inline arwen/host/execution_test.go:1451]: imb ?

File: arwen/host/execution_test.go [hunk 3]
```diff
@@ -1486,77 +1486,6 @@ func TestExecution_Mocked_Wasmer_Instances(t *testing.T) {
 	require.Equal(t, expectedVMOutput, vmOutput)
 }
 
-func TestExecution_GasUsed_SingleContract(t *testing.T) {
```
Comment (sasurobert) [inline arwen/host/execution_test.go:1489]: why did you delete these tests ?

File: arwen/host/execution_vmoutputs_test.go [hunk 4]
```diff
@@ -478,7 +478,7 @@ func expectedVMOutputDestCtxBuiltinDoSomething(input *vmcommon.ContractCallInput
 		nil,
 	)
 	account.Balance = big.NewInt(1000)
-	account.BalanceDelta = big.NewInt(0)
+	account.BalanceDelta = big.NewInt(0).Sub(arwen.One, arwen.One)
```
Comment (sasurobert) [inline arwen/host/execution_vmoutputs_test.go:481]: one minus one is zero.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/contexts/metering.go [hunk 5]
```diff
@@ -99,20 +105,214 @@ func (context *meteringContext) PopDiscard() {
...
+	return clone
+}
+
+func (context *meteringContext) addToGasUsedByAccounts(gasUsed map[string]uint64) {
+	for address, gas := range gasUsed {
+		_, exists := context.gasUsedByAccounts[address]
+		if !exists {
+			context.gasUsedByAccounts[address] = gas
+		} else {
+			context.gasUsedByAccounts[address] += gas
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:140]: Deleted `if` entirely.

File: arwen/contexts/metering.go [hunk 6]
```diff
@@ -99,20 +105,214 @@ func (context *meteringContext) PopDiscard() {
...
+	}
+
+	return gasUsed
+}
+
+func (context *meteringContext) setGasUsedToOutputAccounts(vmOutput *vmcommon.VMOutput) error {
+	for address, account := range vmOutput.OutputAccounts {
+		gasUsed, exists := context.gasUsedByAccounts[address]
+		if exists {
+			account.GasUsed = gasUsed
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:295]: Deleted `if` entirely.

File: arwen/contexts/metering.go [hunk 7]
```diff
@@ -263,6 +390,13 @@ func (context *meteringContext) GasSpentByContract() uint64 {
 	return gasSpent
 }
 
+// GasUsedForExecution returns the actual gas used for execution for the contract which needs to be restored
+func (context *meteringContext) GasUsedForExecution() uint64 {
+	gasUsed := context.GasSpentByContract()
+	gasUsed, _ = math.SubUint64(gasUsed, context.initialCost)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:396]: Yes, the remainder likely isn't used anymore. Most of the usages have been changed on this branch. Will check everywhere.

File: arwen/contexts/output.go [hunk 8]
```diff
@@ -409,51 +396,18 @@ func (context *outputContext) AddTxValueToAccount(address []byte, value *big.Int
...
-
-	context.removeNonUpdatedCode(context.outputState)
-
-	err := context.checkGas(remainedFromForwarded)
+	err := metering.UpdateGasStateOnSuccess(context.outputState)
 	if err != nil {
 		return context.CreateVMOutputInCaseOfError(err)
 	}
 
+	context.removeNonUpdatedCode()
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/output.go:410]: Correct. Moved at the top of method.

File: arwen/host/execution.go (hunk 1 above)
Comment (camilbancioiu) [reply arwen/host/execution.go:903]: Reverted. Accidentally forgot to uncomment it...

File: arwen/host/execution_test.go (hunk 2 above)
Comment (camilbancioiu) [reply arwen/host/execution_test.go:1442]: Fixed.

File: arwen/host/execution_test.go (hunk 3 above)
Comment (camilbancioiu) [reply arwen/host/execution_test.go:1489]: Not deleted, just moved them to `execution_gas_test.go`, and added some more.

File: arwen/host/execution_vmoutputs_test.go (hunk 4 above)
Comment (camilbancioiu) [reply arwen/host/execution_vmoutputs_test.go:481]: This should not be needed, but `big.Int` is bugged. The following test fails:
```
func TestZeroEquality(t *testing.T) {
	zero := big.NewInt(0)
	one := big.NewInt(1)
	zERO := big.NewInt(0).Sub(one, one)
	require.Equal(t, zero, zERO)
}
```

Comment (camilbancioiu) [reply arwen/host/execution_vmoutputs_test.go:481]: Cannot easily use `big.Int.Cmp()` because `account.BalanceDelta` is part of a larger `VMOutput`.

File: arwen/host/execution.go [hunk 9]
```diff
@@ -199,11 +199,55 @@ func copyTxHashesFromContext(copyEnabled bool, runtime arwen.RuntimeContext, inp
...
+func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (vmOutput *vmcommon.VMOutput, asyncInfo *arwen.AsyncContextInfo, err error) {
 	log.Trace("ExecuteOnDestContext", "caller", input.CallerAddr, "dest", input.RecipientAddr, "function", input.Function)
 
-	bigInt, _, metering, output, runtime, storage := host.GetContexts()
+	_, _, _, output, _, _ := host.GetContexts()
+
+	builtinFunctionExecuted := false
+	var builtinRevertInput *vmcommon.ContractCallInput
+
+	// TODO refactor/simplify this block
```
Comment (sasurobert) [inline arwen/host/execution.go:210]: let's refactor this. take the built in function execution into a separate.

Comment (sasurobert) [reply arwen/host/execution.go:210]: transferAndExecute model should be used

File: integrationTests/json/vm_test.go [hunk 10]
```diff
@@ -82,10 +82,12 @@ func TestRustPayableFeaturesLatest(t *testing.T) {
 }
 
 func TestRustAsyncCalls(t *testing.T) {
+	// TODO fix excluded tests and include them back
```
Comment (sasurobert) [inline integrationTests/json/vm_test.go:85]: this TODO still remains ?

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/contexts/metering.go [hunk 11]
```diff
@@ -56,10 +50,20 @@ func NewMeteringContext(
...
+	context.gasUsedByAccounts = make(map[string]uint64)
+}
+
+// InitStateFromContractCallInput initializes the internal state of the
+// MeteringContext using values taken from the provided ContractCallInput
+func (context *meteringContext) InitStateFromContractCallInput(input *vmcommon.VMInput) {
+	context.InitState()
+	context.unlockGasIfAsyncCallback(input)
+	context.initialGasProvided = input.GasProvided
+	context.gasForExecution = input.GasProvided
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:66]: `input.GasProvided` is used here by both fields only as initialization. `context.gasForExecution` is quickly overwritten by `context.deductInitialGas()`. It needs an initial value during isolated tests though.

File: arwen/contexts/metering.go [hunk 12]
```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
...
+	for address, gas := range gasUsed {
+		context.gasUsedByAccounts[address] += gas
+	}
+}
+
+func (context *meteringContext) UpdateGasStateOnSuccess(vmOutput *vmcommon.VMOutput) error {
+	context.updateSCGasUsed()
+	context.setGasUsedToOutputAccounts(vmOutput)
+
+	err := context.checkGasLegacy()
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:144]: `context.checkGasLegacy()` is a no-op if `host.IsArwenV2Enabled()`. It's only here for backwards compatibility, because the actual `context.checkGas()` will report errors for older transactions, before gas forwarding and the gas invariant were introduced.

File: arwen/contexts/metering.go [hunk 13]
```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
...
+	return clone
+}
+
+func (context *meteringContext) addToGasUsedByAccounts(gasUsed map[string]uint64) {
+	for address, gas := range gasUsed {
+		context.gasUsedByAccounts[address] += gas
+	}
+}
+
+func (context *meteringContext) UpdateGasStateOnSuccess(vmOutput *vmcommon.VMOutput) error {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:140]: Added.

File: arwen/contexts/metering.go [hunk 14]
```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
...
+
+	err = context.checkGas(vmOutput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (context *meteringContext) UpdateGasStateOnFailure(vmOutput *vmcommon.VMOutput) {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:157]: Added.

File: arwen/contexts/metering.go [hunk 15]
```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
...
+	callerContractAddress := string(builtinInput.CallerAddr)
+	outputAccounts := builtinOutput.OutputAccounts
+	callerContractAccount, exists := outputAccounts[callerContractAddress]
+
+	gasTransferredByCurrentAccount := uint64(0)
+	if exists {
+		gasTransferredByCurrentAccount = context.getGasTransferredByAccount(callerContractAccount)
+	}
+
+	gasUsedByOthers := context.getGasUsedByAllOtherAccounts(outputAccounts)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:198]: At the moment, this isn't actually needed, because the builtin functions that we currently have are not setting `outputAccount.GasUsed` in `VMOutput`. 

But if we ever have ideas about new builtin functions that will set `outputAccount.GasUsed`, the code is already here and doesn't need to be added.

Also, it's for consistency: builtin functions return a generic `VMOutput`, and this function treats it just like a generic `VMOutput`, without assumptions (including handling output accounts with `GasUsed`).

File: arwen/contexts/metering.go [hunk 16]
```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
...
+
+	gasUsed := math.SubUint64(builtinInput.GasProvided, builtinOutput.GasRemaining)
+	gasUsed = math.SubUint64(gasUsed, gasTransferredByCurrentAccount)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:201]: That would introduce assumptions about the behavior of future builtin functions. See comment above.

File: arwen/contexts/metering.go [hunk 17]
```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
...
+
+func (context *meteringContext) getGasUsedByAllOtherAccounts(outputAccounts map[string]*vmcommon.OutputAccount) uint64 {
+
+	gasUsedAndTransferred := uint64(0)
+	currentAccountAddress := string(context.host.Runtime().GetSCAddress())
+	for address, account := range outputAccounts {
+		gasTransferred := context.getGasTransferredByAccount(account)
+
+		gasUsed := uint64(0)
+		if address != currentAccountAddress {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:265]: When this code is executed, any child called by the current contract has already finished its execution. But it also means that all the gas spent on child executions is still found in `metering.GasSpentByContract()`.

So the gas used by the current account is `metering.GasSpentByContract()` minus the entire sum produced by `getGasUsedByAllOtherAccounts()`.

File: arwen/contexts/output.go [hunk 18]
```diff
@@ -164,6 +149,10 @@ func (context *outputContext) GetOutputAccount(address []byte) (*vmcommon.Output
 	return account, accountIsNew
 }
 
+func (context *outputContext) GetOutputAccounts() map[string]*vmcommon.OutputAccount {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/output.go:152]: Added.

File: arwen/host/asyncCall.go [hunk 19]
```diff
@@ -37,10 +37,7 @@ func (host *vmHost) handleAsyncCallBreakpoint() error {
 	// Cross-shard calls for built-in functions must be executed in both the
 	// sender and destination shards.
 	if execMode == arwen.AsyncBuiltinFuncCrossShard {
-		vmOutput, gasUsedBeforeReset, err := host.executeSyncDestinationCall(asyncCallInfo)
-		if err == nil && vmOutput.ReturnCode == vmcommon.Ok {
-			host.meteringContext.UseGas(gasUsedBeforeReset)
-		}
+		vmOutput, err := host.executeSyncDestinationCall(asyncCallInfo)
 		if vmOutput != nil {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:41]: Correct, removed.

File: arwen/host/asyncCall.go [hunk 20]
```diff
@@ -138,7 +135,7 @@ func (host *vmHost) executeSyncDestinationCall(asyncCallInfo arwen.AsyncCallInfo
 		"func", destinationCallInput.Function,
 		"args", destinationCallInput.Arguments)
 
-	destinationVMOutput, _, gasUsedBeforeReset, err := host.ExecuteOnDestContext(destinationCallInput)
+	destinationVMOutput, _, err := host.ExecuteOnDestContext(destinationCallInput)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:139]: Done and deleted `if destinationVMOutput != nil`.

File: arwen/host/asyncCall.go [hunk 21]
```diff
@@ -180,16 +172,7 @@ func (host *vmHost) executeSyncCallbackCall(
...
-	// used points should be reset before actually entering the callback execution
-	host.runtimeContext.SetPointsUsed(0)
-	callbackVMOutput, _, _, callBackErr := host.ExecuteOnDestContext(callbackCallInput)
-
-	noErrorOnCallback := callBackErr == nil && callbackVMOutput.ReturnCode == vmcommon.Ok
-	noErrorOnAsyncCall := destinationErr == nil && destinationVMOutput.ReturnCode == vmcommon.Ok
-	if noErrorOnCallback && noErrorOnAsyncCall && execMode != arwen.AsyncBuiltinFuncIntraShard {
-		host.meteringContext.UseGas(gasConsumedForExecution)
-	}
+	callbackVMOutput, _, callBackErr := host.ExecuteOnDestContext(callbackCallInput)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:180]: Done and deleted `if callbackVMOutput != nil`.

File: integrationTests/json/vm_test.go (hunk 10 above)
Comment (camilbancioiu) [reply integrationTests/json/vm_test.go:85]: Still remains, but only for some of the excluded tests. I have restored 3 of them, because now they pass.

File: arwen/contexts/metering.go [hunk 22]
```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
...
+	gasUsedByOthers := context.getGasUsedByAllOtherAccounts(outputAccounts)
+
+	gasUsed := context.GasSpentByContract()
+	gasUsed = math.SubUint64(gasUsed, gasTransferredByCurrentAccount)
+	gasUsed = math.SubUint64(gasUsed, gasUsedByOthers)
+
+	context.gasUsedByAccounts[string(currentAccountAddress)] = gasUsed
+}
+
+func (context *meteringContext) TrackGasUsedByBuiltinFunction(err error, builtinInput *vmcommon.ContractCallInput, builtinOutput *vmcommon.VMOutput) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/metering.go:190]: isn't this duplicated ?

File: arwen/contexts/metering.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/metering.go]: we do not care about legacy as we will resolve this from the factory.

File: arwen/host/execution.go (hunk 9 above)
Comment (camilbancioiu) [reply arwen/host/execution.go]: Refactored.

File: arwen/host/execution_gas_test.go [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+			input.RecipientAddr = arguments[callIndex+0]
+			input.Function = string(arguments[callIndex+1])
+			numCalls := big.NewInt(0).SetBytes(arguments[callIndex+2]).Uint64()
+
+			for i := uint64(0); i < numCalls; i++ {
+				_, _, err := host.ExecuteOnDestContext(input)
+				require.Nil(t, err)
+			}
+		}
+	})
```
Comment (alyn509) [inline arwen/host/execution_gas_test.go:251]: These 2 have quite some similarities.

I think we can reduce some duplicity here.
Knowing argsPerCall we can make sure that `for callIndex := 0; callIndex < len(arguments); callIndex += argsPerCall ` loops only once if execOnSameCtx

About the error taken from **ExecuteOnSameContext or **ExecuteOnDestContext** we can handle that from function parameter

**Review by matei-p — APPROVED** [review]

File: arwen/contexts/metering.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go]: Added extra logic dedicated to SC execution after builtin calls.

File: arwen/host/execution.go [hunk 24]
```diff
@@ -199,11 +199,57 @@ func copyTxHashesFromContext(copyEnabled bool, runtime arwen.RuntimeContext, inp
...
+		builtinFunctionExecuted = true
+		input, vmOutput, err = host.handleBuiltinFunctionCall(input)
+		if err != nil {
+			vmOutput = host.Output().CreateVMOutputInCaseOfError(err)
+			return
+		}
+	}
+
+	if input != nil {
+		vmOutput, asyncInfo, err = host.executeOnDestContextNoBuiltinFunction(input)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:218]: When the original input contains a call to a SC method, but wrapped in an `ESDTTransfer`. This "wrapping" is used to send ESDT to a contract *and* call one of its methods. During the call, the contract will see that it has received the aforementioned tokens as part of the call itself.

File: arwen/host/execution_gas_test.go (hunk 23 above)
Comment (camilbancioiu) [reply arwen/host/execution_gas_test.go:251]: Refactoring these tests is addressed by the branch `gas-async-testing`. But you're right, these can be simplified.

File: arwen/contexts/metering.go [hunk 25]
```diff
@@ -99,20 +105,209 @@ func (context *meteringContext) PopDiscard() {
...
+func (context *meteringContext) getCurrentTotalUsedGas() uint64 {
+	outputAccounts := context.host.Output().GetOutputAccounts()
+
+	gasUsed := uint64(0)
+	for _, outputAccount := range outputAccounts {
+		gasUsed = math.AddUint64(gasUsed, outputAccount.GasUsed)
+		for _, outputTransfer := range outputAccount.OutputTransfers {
+			gasUsed = math.AddUint64(gasUsed, outputTransfer.GasLimit)
+			gasUsed = math.AddUint64(gasUsed, outputTransfer.GasLocked)
+		}
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go:259]: Indeed, refactored.

**Review by matei-p — DISMISSED** [review]

File: integrationTests/json/vm_test.go [hunk 26]
```diff
@@ -86,7 +86,6 @@ func TestRustAsyncCalls(t *testing.T) {
 	runTestsInFolder(t, "features/async/mandos", []string{
 		"features/async/mandos/recursive_caller_esdt_2.scen.json",
 		"features/async/mandos/recursive_caller_esdt_x.scen.json",
```
Comment (sasurobert) [inline integrationTests/json/vm_test.go:88]: you actually deleted the test

File: test/contracts/async-call-child/elrond.json [hunk 27]
```diff
@@ -0,0 +1,6 @@
+{
+    "language": "clang",
+    "source_files": [
+        "/home/camil.bancioiu/Work/Elrond/arwen-wasm-vm/test/contracts/async-call-child/async-call-child.c"
```
Comment (sasurobert) [inline test/contracts/async-call-child/elrond.json:4]: is this link good ? i think this should be relative path

File: arwen/contexts/metering.go [hunk 28]
```diff
@@ -193,20 +193,7 @@ func (context *meteringContext) TrackGasUsedByBuiltinFunction(err error, builtin
 		return err
```
Comment (sasurobert) [inline arwen/contexts/metering.go:198]: it is no sense to return the error back. just send the error and that's it.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: integrationTests/json/vm_test.go [hunk 29]
```diff
@@ -82,9 +82,8 @@ func TestRustPayableFeaturesLatest(t *testing.T) {
 }
 
 func TestRustAsyncCalls(t *testing.T) {
+	// TODO fix excluded tests and include them back
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply integrationTests/json/vm_test.go:85]: This is an exclusion list. After more debugging, the remaining two tests in this list will be deleted as well (i.e. executed normally). Will do in a separate bugfixing PR.

File: arwen/host/execution.go [hunk 30]
```diff
@@ -806,10 +776,11 @@ func (host *vmHost) callBuiltinFunction(input *vmcommon.ContractCallInput) (*vmc
 		}
 	}
 
+	err = metering.TrackGasUsedByBuiltinFunction(err, input, vmOutput, newVMInput)
+
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:780]: Removed the returned error entirely, it wasn't actually needed (was forgotten here from previous code).

File: arwen/host/execution.go [hunk 31]
```diff
@@ -764,40 +755,19 @@ func (host *vmHost) ExecuteESDTTransfer(destination []byte, sender []byte, token
...
-	if vmOutput.GasRemaining < input.GasProvided {
-		metering.UseGas(gasConsumed)
+		err = metering.TrackGasUsedByBuiltinFunction(err, input, vmOutput, nil)
+		return nil, nil, err
 	}
 
 	newVMInput, err := host.isSCExecutionAfterBuiltInFunc(input, vmOutput)
 	if err != nil {
-		return nil, gasConsumedForExecution, err
+		err = metering.TrackGasUsedByBuiltinFunction(err, input, vmOutput, nil)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:769]: Changed `metering.TrackGasUsedByBuiltinFunction()` to not return anything (not needed).

File: arwen/host/execution.go [hunk 32]
```diff
@@ -199,11 +199,57 @@ func copyTxHashesFromContext(copyEnabled bool, runtime arwen.RuntimeContext, inp
...
+	if err != nil {
+		log.Trace("ExecuteOnDestContext builtin function", "error", err)
+		return nil, nil, err
+	}
+
+	output.AddToActiveState(builtinOutput)
+
+	if postBuiltinInput == nil {
+		if input.CallType != vmcommon.AsynchronousCallBack || input.CallValue.Cmp(arwen.Zero) == 0 {
+			err = output.TransferValueOnly(input.RecipientAddr, input.CallerAddr, input.CallValue, false)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:239]: This is a mistake. Deleted.

File: arwen/host/execution.go [hunk 33]
```diff
@@ -199,11 +199,57 @@ func copyTxHashesFromContext(copyEnabled bool, runtime arwen.RuntimeContext, inp
...
+func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (vmOutput *vmcommon.VMOutput, asyncInfo *arwen.AsyncContextInfo, err error) {
 	log.Trace("ExecuteOnDestContext", "caller", input.CallerAddr, "dest", input.RecipientAddr, "function", input.Function)
 
-	bigInt, _, metering, output, runtime, storage := host.GetContexts()
+	builtinFunctionExecuted := false
+	originalInput := input
 
+	if host.IsBuiltinFunctionName(input.Function) {
+		builtinFunctionExecuted = true
+		input, vmOutput, err = host.handleBuiltinFunctionCall(input)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:210]: Added variable `scExecutionInput`.

File: arwen/host/asyncCall.go [hunk 34]
```diff
@@ -37,27 +37,28 @@ func (host *vmHost) handleAsyncCallBreakpoint() error {
...
-			host.meteringContext.UseGas(gasUsedBeforeReset)
-		}
-		if vmOutput != nil {
-			log.LogIfError(err, "async call failed: sync built-in", "error", err,
-				"retCode", vmOutput.ReturnCode,
-				"message", vmOutput.ReturnMessage)
-		}
+		vmOutput, err := host.executeSyncDestinationCall(asyncCallInfo)
+		log.LogIfError(err, "async call failed: sync built-in", "error", err,
+			"retCode", vmOutput.ReturnCode,
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/asyncCall.go:42]: True in general, but in this case, `host.executeSyncDestinationCall()` is guaranteed to return a valid `VMOutput` due to `host.ExecuteOnDestContext()` inside, even if `err != nil`.

File: arwen/contexts/metering.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/metering.go]: Deleted.

File: arwen/contexts/metering.go (hunk 28 above)
Comment (camilbancioiu) [reply arwen/contexts/metering.go:198]: Removed `return err` and added a `log.Trace()`.

File: test/contracts/async-call-child/elrond.json (hunk 27 above)
Comment (camilbancioiu) [reply test/contracts/async-call-child/elrond.json:4]: Not good. Fixed.

**Review by matei-p — DISMISSED** [review]

File: arwen/contexts/metering.go [hunk 35]
```diff
@@ -99,20 +105,187 @@ func (context *meteringContext) PopDiscard() {
...
+	gasUsedByOthers := context.getGasUsedByAllOtherAccounts(outputAccounts)
+
+	gasUsed := context.GasSpentByContract()
+	gasUsed = math.SubUint64(gasUsed, gasTransferredByCurrentAccount)
+	gasUsed = math.SubUint64(gasUsed, gasUsedByOthers)
+
+	context.gasUsedByAccounts[string(currentAccountAddress)] = gasUsed
+}
+
+func (context *meteringContext) TrackGasUsedByBuiltinFunction(
```
Comment (sasurobert) [inline arwen/contexts/metering.go:185]: comment on public method. run a linter.

File: arwen/contexts/metering.go [hunk 36]
```diff
@@ -99,20 +105,187 @@ func (context *meteringContext) PopDiscard() {
...
+	err error,
+	builtinInput *vmcommon.ContractCallInput,
+	builtinOutput *vmcommon.VMOutput,
+	postBuiltinInput *vmcommon.ContractCallInput,
+) {
+	if err != nil {
```
Comment (sasurobert) [inline arwen/contexts/metering.go:191]: I think this could be called without the error input. and copy simply the useGas function wherever is needed. Only 3 places as I saw.

**Review by sasurobert — CHANGES_REQUESTED** [review]

**Review by matei-p — DISMISSED** [review]

File: arwen/contexts/metering.go (hunk 36 above)
Comment (camilbancioiu) [reply arwen/contexts/metering.go:191]: Agreed, changed accordingly.

**Review by sasurobert — APPROVED** [review]

**Review by matei-p — APPROVED** [review]
